torchtune/modules/early_exit_loss.py

```diff
@@ -0,0 +1,334 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import copy
+from enum import Enum
+from typing import Dict, List, Optional
+
+import numpy as np
+import torch
+
+from torchtune import utils
+from torchtune.modules.transformer import TransformerDecoder
+
+log = utils.get_logger("DEBUG")
+
+
+class LossScaleType(str, Enum):
```

Comment:
General question about some of these enums (e.g. this one, `EarlyExitCurriculumType`, `ScaleType`). Do you find that it is easier to use these rather than just have e.g. `layer_ids_to_loss_scales` be an arbitrary `Callable[[torch.Tensor, int], torch.Tensor]` (i.e. that is taking in `layer_ids` and `n_layers` and returning the scales)? If so, why? Personally I feel it's a bit more flexible to do it the latter way but would be interested to know if you find the Enum approach to be better for experimentation (or for other reasons I'm missing). Anyways not a request for a change, just wanna hear your thoughts.

Reply:
> I am trying to look for an example to follow but can't find. For example, how can layer_ids_to_loss_scales determine the scale type?

Basically I am thinking of something like this.

In the `early_exit_loss` function:

```
def early_exit_loss(
	# keep all the existing args except loss_scale_type, replace it with..
	loss_scale_fn: Callable[[Tensor, int, float], Tensor]
):
	# Everything in here is the same except the call to layer_ids_to_loss_scales becomes..
	losses_scales = loss_scale_fn(
	        torch.Tensor(hidden_layer_ids).to(losses_early),
	        len(model.layers),
		e_scale,
	)
```

Then we define all the different loss scale types like this:

```
def inv_sqrt_l_loss_scale(layer_ids: Tensor, n_layers: int, e_scale: float = 1.0) -> Tensor:
	loss_scales = torch.reciprocal(torch.sqrt(layer_ids + 1))
	loss_scales = loss_scales * torch.where(layer_ids < n_layers - 1, e_scale, 1.0)
	return loss_scales / torch.sum(loss_scales)
```

And finally the config would look like:

```
early_exit_loss:
  layers: ":"
  curriculum: "gradual"
  scale_type: torchtune.modules.early_exit_loss.inv_sqrt_l_loss_scale
  scale: 1.0
```

Note that the one final thing needed would be to convert the `scale_type` import path to the actual function it represents (since OmegaConf will just see this as a string). However, we have [_get_component_from_path](https://github.com/pytorch/torchtune/blob/e9b9ea568e169aea13ae2d03c0eb92ee241f47a1/torchtune/config/_utils.py#L37) for exactly this case.

The reason I like this approach is that then if I am a user who wants to experiment with my own scaling method, I just need to write the Python function for it and point the config to its path on my local machine. No need to go in and modify the enum.

> In test_layer_ids_to_loss_scales, I just made the PyTest fixture test all the possible values in the enum:

Oh that is quite nice. I guess we could define a global variable containing all the different supported functions, but I guess that's pretty much just a worse version of an enum 😅

Reply:
Changed scale_tyoe from enum to function in 00d8efa7ca63f6c10db0fc7ace1143f8f6ca71c1 and changed curriculum from enum to class in 78b8996545149433a69c1ba5a991c9a596c60708.